contracts/extensions/bridge-operator-governance/BridgeManager.sol

```diff
@@ -0,0 +1,511 @@
+  bytes32 private constant GOVERNOR_TO_BRIDGE_OPERATOR_INFO_SLOT =
+    0x88547008e60f5748911f2e59feb3093b7e4c2e87b2dd69d61f112fcc932de8e3;
+  /// @dev value is equal to keccak256("@ronin.dpos.gateway.BridgeAdmin.govenorOf.slot") - 1
+  bytes32 private constant GOVENOR_OF_SLOT = 0x8400683eb2cb350596d73644c0c89fe45f108600003457374f4ab3e87b4f3aa3;
+  /// @dev value is equal to keccak256("@ronin.dpos.gateway.BridgeAdmin.governors.slot") - 1
+  bytes32 private constant GOVERNOR_SET_SLOT = 0x546f6b46ab35b030b6816596b352aef78857377176c8b24baa2046a62cf1998c;
+  /// @dev value is equal to keccak256("@ronin.dpos.gateway.BridgeAdmin.bridgeOperators.slot") - 1
+  bytes32 private constant BRIDGE_OPERATOR_SET_SLOT =
+    0xd38c234075fde25875da8a6b7e36b58b86681d483271a99eeeee1d78e258a24d;
+
+  /**
+   * @dev The domain separator used for computing hash digests in the contract.
+   */
+  bytes32 public immutable DOMAIN_SEPARATOR;
+
+  /**
+   * @dev The numerator value used for calculations in the contract.
+   */
+  uint256 internal _num;
+
+  /**
+   * @dev The denominator value used for calculations in the contract.
+   */
+  uint256 internal _denom;
+
+  /**
+   * @dev The nonce value used for tracking nonces in the contract.
+   */
+  uint256 internal _nonce;
+
+  /**
+   * @dev The total weight value used for storing the cumulative weight in the contract.
+   */
+  uint256 internal _totalWeight;
+
+  modifier onlySelfCall() {
+    _requireSelfCall();
+    _;
+  }
+
+  /**
+   * @dev Modifier to ensure that the elements in the `arr` array are non-duplicates.
+   * It calls the internal `_checkDuplicate` function to perform the duplicate check.
+   *
+   * Requirements:
+   * - The elements in the `arr` array must not contain any duplicates.
+   */
+  modifier nonDuplicate(address[] memory arr) {
+    _checkDuplicate(arr);
+    _;
+  }
+
+  constructor(
+    uint256 num,
+    uint256 denom,
+    uint256 roninChainId,
+    address admin,
+    uint256[] memory voteWeights,
+    address[] memory governors,
+    address[] memory bridgeOperators
+  ) payable nonDuplicate(governors) nonDuplicate(bridgeOperators) {
+    _checkNonZeroAddress(admin);
+    assembly {
+      sstore(_ADMIN_SLOT, admin)
+    }
+
+    _nonce = 1;
+    _num = num;
+    _denom = denom;
+
+    DOMAIN_SEPARATOR = keccak256(
+      abi.encode(
+        keccak256("EIP712Domain(string name,string version,bytes32 salt)"),
+        keccak256("BridgeAdmin"), // name hash
+        keccak256("2"), // version hash
+        keccak256(abi.encode("BRIDGE_ADMIN", roninChainId)) // salt
+      )
+    );
+
+    _addBridgeOperators(voteWeights, governors, bridgeOperators);
+  }
+
+  /**
+   * @inheritdoc IBridgeManager
+   */
+  function addBridgeOperators(
+    uint256[] calldata voteWeights,
+    address[] calldata governors,
+    address[] calldata bridgeOperators
+  ) external onlySelfCall nonDuplicate(bridgeOperators) nonDuplicate(governors) returns (bool[] memory addeds) {
```

Comment:
Why not `nonDuplicate([...bridgeOperators, ...governors])`? I think we should restrict `bridgeOperator != governor` to avoid misuse by bridge operators.

Reply:
Are their any efficient ways in implementing concatenations?

Reply:
updated.